### PR TITLE
MGMT-13888: Remove spoke node on BMH delete

### PR DESCRIPTION
**Pull request by carbonin**
This PR builds on https://github.com/openshift/assisted-service/pull/4948 by cleaning and removing the node from the spoke cluster when the BMH is annotated and deleted.

This allows ZTP users (specifically ones using git-ops where they don't have the ability to alter/remove the agent CR) to deprovision nodes using the same mechanisms they used to originally provision them (the BMH).

This does not handle cordoning and draining the spoke node before removing it. That will be handled in a follow up (https://issues.redhat.com/browse/MGMT-13890)

This requires the following fixes in BMO:
https://github.com/metal3-io/baremetal-operator/pull/1184
https://github.com/metal3-io/baremetal-operator/pull/1230

This means that this will only work with 4.13 nightlies after this merges.

## How it works

- When the user annotates the BMH and the converged flow is in use a finalizer is added to the BMH.
- When the BMH is deleted this finalizer (and the work in ~~https://github.com/metal3-io/baremetal-operator/pull/1230~~ https://github.com/metal3-io/baremetal-operator/pull/1236) allows BMAC time to reattach and deprovision the BMH. Once these edits are made BMAC removes our finalizer.
- BMO then boots the host back into the preprovisioning image which runs the ironic agent and cleans the host (this is required to ensure that the host doesn't continue running OCP if/when it is rebooted after it is removed).
- Throughout this process the agent controller has been waiting for the BMH to be fully deleted as a part of its finalizer logic.
- After BMO finishes deprovisioning the host it removes its finalizer on the BMH and allows it to be garbage collected.
- Once the BMH is gone the agent controller contacts the spoke cluster and removes the node record.

## List all the issues related to this PR

https://issues.redhat.com/browse/MGMT-13888
https://issues.redhat.com/browse/RFE-3431

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

Tested manually on the dev-scripts cluster with an edited BMO.

1. Install a spoke using ZTP
2. Annotate the BMHs with `bmac.agent-install.openshift.io/remove-agent-and-node-on-delete: true`
3. Delete a BMH
4. Observe BMH deprovision and node removal

- [ ] assisted-test-infra environment
- [x] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc) **TODO**
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by carbonin** [comment]
Left this as a draft for a few reasons:

1. https://github.com/metal3-io/baremetal-operator/pull/1230 has not made it into the downstream BMO code yet
2. agent controller changes still need tests
3. I'm still looking into some errors I see sometimes when deprovisioning a bunch of hosts at once

1 and 2 should be easy enough to deal with, but 3 requires some more work.

Specifically it seems like when BMO can't push its updates to the BMH it moves the host straight to deleting rather than allowing it to deprovision. I thought this was what @dtantsur's [fix](https://github.com/metal3-io/baremetal-operator/pull/1184) handled, but maybe there's a few cases that still need looking into.

The BMO logs for this case for me look like this in case anyone's interested:
```
{"level":"error","ts":"2023-03-10T21:10:21.174Z","msg":"Reconciler error","controller":"baremetalhost","controllerGroup":"metal3.io","controllerKind":"BareMetalHost","BareMetalHost":{"name":"ostest-extraworker-0","namespace":"assisted-installer"},"namespace":"assisted-installer","name":"ostest-extraworker-0","reconcileID":"b7e0d462-2e5d-43a9-a3e3-76275b0f24b3","error":"failed to save host status after \"provisioned\": Operation cannot be fulfilled on baremetalhosts.metal3.io \"ostest-extraworker-0\": the object has been modified; please apply your changes to the latest version and try again","errorVerbose":"Operation cannot be fulfilled on baremetalhosts.metal3.io \"ostest-extraworker-0\": the object has been modified; please apply your changes to the latest version and try again\nfailed to save host status after \"provisioned\"\ngithub.com/metal3-io/baremetal-operator/controllers/metal3%2eio.(*BareMetalHostReconciler).Reconcile\n\t/go/src/github.com/metal3-io/baremetal-operator/controllers/metal3.io/baremetalhost_controller.go:246\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/go/src/github.com/metal3-io/baremetal-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/src/github.com/metal3-io/baremetal-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/metal3-io/baremetal-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/src/github.com/metal3-io/baremetal-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234\nruntime.goexit\n\t/usr/lib/golang/src/runtime/asm_amd64.s:1594","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/src/github.com/metal3-io/baremetal-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/src/github.com/metal3-io/baremetal-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/src/github.com/metal3-io/baremetal-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234"}
{"level":"info","ts":"2023-03-10T21:10:21.174Z","logger":"controllers.BareMetalHost","msg":"start","baremetalhost":"assisted-installer/ostest-extraworker-0"}
{"level":"info","ts":"2023-03-10T21:10:21.174Z","logger":"controllers.BareMetalHost","msg":"hardwareData is ready to be deleted","baremetalhost":"assisted-installer/ostest-extraworker-0"}
2023-03-10T21:10:21.174Z        DEBUG   provisioner.ironic      verifying ironic provisioner dependencies       {"host": "assisted-installer~ostest-extraworker-0"}
...
{"level":"info","ts":"2023-03-10T21:10:21.211Z","logger":"controllers.BareMetalHost","msg":"Initiating host deletion","baremetalhost":"assisted-installer/ostest-extraworker-0","provisioningState":"provisioned"}
{"level":"info","ts":"2023-03-10T21:10:21.211Z","logger":"controllers.BareMetalHost","msg":"changing provisioning state","baremetalhost":"assisted-installer/ostest-extraworker-0","provisioningState":"provisioned","old":"provisioned","new":"deleting"}
{"level":"info","ts":"2023-03-10T21:10:21.211Z","logger":"controllers.BareMetalHost","msg":"saving host status","baremetalhost":"assisted-installer/ostest-extraworker-0","provisioningState":"provisioned","operational status":"detached","provisioning state":"deleting"}
{"level":"info","ts":"2023-03-10T21:10:21.221Z","logger":"controllers.BareMetalHost","msg":"done","baremetalhost":"assisted-installer/ostest-extraworker-0","provisioningState":"provisioned","requeue":true,"after":0}
{"level":"info","ts":"2023-03-10T21:10:21.221Z","logger":"controllers.BareMetalHost","msg":"start","baremetalhost":"assisted-installer/ostest-extraworker-0"}
{"level":"info","ts":"2023-03-10T21:10:21.221Z","logger":"controllers.BareMetalHost","msg":"hardwareData is ready to be deleted","baremetalhost":"assisted-installer/ostest-extraworker-0"}
2023-03-10T21:10:21.221Z        DEBUG   provisioner.ironic      verifying ironic provisioner dependencies       {"host": "assisted-installer~ostest-extraworker-0"}
{"level":"info","ts":"2023-03-10T21:10:21.221Z","logger":"controllers.BMCEventSubscription","msg":"start","bmceventsubscription":"assisted-installer/ostest-extraworker-0"}
{"level":"info","ts":"2023-03-10T21:10:21.221Z","logger":"controllers.BMCEventSubscription","msg":"done","bmceventsubscription":"assisted-installer/ostest-extraworker-0"}
```

**Comment by carbonin** [comment]
cc @filanov @eranco74 @mhrivnak @avishayt @jgato

**Comment by carbonin** [comment]
Another open question is that since I know this won't work with BMO < 4.13 (read: hub version < 4.13) should we add some kind of explicit check for that like we do for converged?

With earlier versions the BMH will be deleted and the agent will go away, but the spoke node won't get deprovisioned (or rather the `Node` will get removed, but it will come right back).

**Comment by mhrivnak** [comment]
Can you elaborate on the difference in behavior and what change causes it in 4.13?

What the check look like that you're doing elsewhere?

File: internal/controller/controllers/agent_controller.go
Comment (eranco74) [inline internal/controller/controllers/agent_controller.go:156]: We should probably delete the BMH from the spoke as well.
Not sure if here or in BMAC

File: internal/controller/controllers/bmh_agent_controller.go
Comment (eranco74) [inline internal/controller/controllers/bmh_agent_controller.go:387]: What's the purpose of this annotation?

Comment (filanov) [reply internal/controller/controllers/bmh_agent_controller.go:387]: if i understand correctly it's a mark for the agent to wait until bmh is deleted first
so it will mark with an annotation that the agent will monitor, should probably add a comment explaining a bit why the annotation is added, otherwise reading the code will need to go to the commit message.

File: internal/controller/controllers/agent_controller.go
Comment (carbonin) [reply internal/controller/controllers/agent_controller.go:147]: Right now it's requeueing every ten seconds.
I think it's good to have just to show we're still doing something even if the cleaning takes a while, but I can make it longer if we'd rather not have all of these messages.

File: internal/controller/controllers/agent_controller.go
Comment (carbonin) [reply internal/controller/controllers/agent_controller.go:153]: Yeah, this is a tough one because the agent and infraenv are being removed there's nowhere to report this kind of thing. It also didn't seem like a recoverable error and we don't really have a way to do multiple retries in something like this without adding to the status.

File: internal/controller/controllers/bmh_agent_controller.go
Comment (carbonin) [reply internal/controller/controllers/bmh_agent_controller.go:387]: Right, this tells the agent to delete the spoke node when the BMH finishes deprovisioning.

I'll add a comment.

File: internal/controller/controllers/agent_controller.go
Comment (carbonin) [reply internal/controller/controllers/agent_controller.go:156]: Should we assume the spoke BMH is not being used?
In my tests the spoke BMHs are being used by a machine in a machineset.
Should we also try to clean that up?

Also isn't @slaviered making the spoke BMH effectively useless? Maybe we just shouldn't be creating them at all in the future?

Maybe this is better in a followup?

**Comment by carbonin** [comment]
@mhrivnak sorry it took so long to respond to this

> Can you elaborate on the difference in behavior and what change causes it in 4.13?

I'm adding [logic](https://github.com/metal3-io/baremetal-operator/pull/1236) to make BMO delay deleting detached BMHs so that we have time to re-attach plus do the cordon/drain.
This is vendored into this PR now.

> What the check look like that you're doing elsewhere?

We check the BMO cluster operator version against a known good version at startup.
https://github.com/openshift/assisted-service/blob/f6803364b07d7318203825cbfc8875e804f9e4d8/internal/controller/controllers/bmo_utils.go#L40-L62

File: internal/controller/controllers/agent_controller.go
Comment (jgato) [reply internal/controller/controllers/agent_controller.go:156]: What is the intention on having BMHs copied into the spokes? Maybe it is part of the installation, I dont know. But mainly there you have the BMCs information, that would be useful for spoke admin to interact with the BMC? 
Anyway, if the BMHs exists on the spoke, I think it should be deleted. If not, for example, I have experienced problems re-adding the node. So, you would want to delete one node, maybe some maintenance, hardware replacement, and after that.. you decide to re-install the host. If the BMH was not deleted, I have found the CSR is not automatically approved. In the same scenario, if I previously I deleted the BMH, you can re-install the host seamless.

**Comment by carbonin** [comment]
This is ready for review, but it still has my fork of BMO vendored in.

https://github.com/metal3-io/baremetal-operator/pull/1236 is merged, but we need to wait for it to be merged into the openshift fork before I can update our go.mod

File: internal/controller/controllers/agent_controller.go
Comment (avishayt) [inline internal/controller/controllers/agent_controller.go:246]: Is the node name guaranteed to match the hostname?

File: internal/controller/controllers/agent_controller.go
Comment (avishayt) [inline internal/controller/controllers/agent_controller.go:243]: Why don't we fail here (and below if DeleteNode fails)?

File: internal/controller/controllers/agent_controller.go
Comment (carbonin) [reply internal/controller/controllers/agent_controller.go:246]: This is the [assumption](https://github.com/openshift/assisted-service/blob/master/docs/enhancements/auto-return-agent-to-infra-env.md#assumptions) I made during the agent reclaim work, so I just used it again here. I haven't heard of issues with this approach.

File: internal/controller/controllers/agent_controller.go
Comment (carbonin) [reply internal/controller/controllers/agent_controller.go:243]: It didn't seem like a recoverable error to me and if we fail we'd need to know when to give up. We'd have to implement some kind of tracker or timeout for retries.

It's also very possible that someone is deleting an SNO or an entire cluster and we won't be able to connect to delete the node.

**Review by avishayt — APPROVED** [review]

**Comment by carbonin** [comment]
/hold

Marking this as ready for review as it's good enough to run through CI, but I want to do one more round of manual testing before merging.

**Comment by carbonin** [comment]
I also depended on BMO's master branch, but if we want to keep it to 4.13 for some reason we'd have to wait for https://github.com/openshift/baremetal-operator/pull/261 to merge

**Comment by carbonin** [comment]
Retested this with the latest changes and BMO with the patch included.

Merging this without the BMO patch available isn't much of a risk as the feature is opt-in with the annotation and even if someone were to add the annotation the host just wouldn't be cleaned.

**Review by avishayt — APPROVED** [review]

**Pull request merged**
